src/libasr/asr_utils.cpp

```diff
@@ -74,7 +74,8 @@ std::vector<std::string> determine_function_definition_order(
             ASR::Function_t* func = ASR::down_cast<ASR::Function_t>(itr.second);
             for( size_t i = 0; i < func->n_dependencies; i++ ) {
                 std::string dep = func->m_dependencies[i];
-                deps.push_back(dep);
+                if (symtab->get_symbol(dep) != nullptr)
+                    deps.push_back(dep);
```

Comment:
I think the "contract" (checked by "verify") is that all the identifiers in dependencies are from the local symbol table, so then `symtab->get_symbol(dep)` should never be null. Can you post an example where this fails? It seems this PR is a workaround for some bug, that we should rather fix directly.

Reply:
In this example:
```py
from ltypes import i32

SIZE: i32 = i32(1)

def main() -> None:
    xs: i32[SIZE]

main()
```

The ASR has:
```console
(SymbolTable
                        2
                        {
                            xs:
                                (Variable 
                                    2 
                                    xs 
                                    [SIZE] 
                                    Local 
                                    () 
                                    () 
                                    Default 
                                    (Integer 4 [((IntegerConstant 0 (Integer 4 [])) 
                                    (Var 1 SIZE))]) 
                                    Source 
                                    Public 
                                    Required 
                                    .false.
                                )
                            
                        }) 
```

We see that `xs` has `SIZE` as the dependent variable which is present in the parent symbol.

Reply:
I guess, having `SIZE` in dependencies is correct, after all it depends on it. Yes, but including it in declaration order doesn't make sense, because parent scope must have been already visited. I guess a better way would be to include `SIZE` in `main`'s (the function) dependency list, because we want to make sure `SIZE` is declared before `main`. What do you say @certik ?

Reply:
I see, thanks @Smit-create for the example. Now it make sense what is going on.

What is the best design?

The motivation for these dependencies is to ensure we can create the symbol table in the right order. Yes, things can depend on parent scope, but as Gagandeep said, the parent scope is implicitly always created first anyway.

One option is to treat dependencies as strictly "local" dependencies, only within the given symbol table. That is how it is used in the backend.  We can rename the argument to "local_dependencies". Any dependency on parent scope is implicit.

Then in the parent scope, there is also a dependency: `main` depends on `SIZE`.

How to easily figure out these dependencies? Not sure.

Another approach is to always store all dependencies, but even then we still need to add SIZE to main in the parent scope, otherwise main() could have been generated before SIZE.

There is another issue. One can have two mutually recursive functions that call each other. That is the reason why we need to do a two pass approach in AST->ASR. So they will both depend on each other. I guess that could still be encoded as a cyclic dependency.

Reply:
I think we can only include the local dependencies with the given symbol table. What do you think @czgdp1807 ?

Reply:
I would only use dependencies for local symbols.